rtcdatachannel.go

```diff
@@ -89,24 +89,61 @@ type RTCDataChannel struct {
 	// OnError             func()
 	// OnClose             func()
 
-	// Onmessage designates an event handler which is invoked on a message
-	// arrival over the sctp transport from a remote peer.
-	//
-	// Deprecated: use OnMessage instead.
-	Onmessage func(datachannel.Payload)
-
-	// OnMessage designates an event handler which is invoked on a message
-	// arrival over the sctp transport from a remote peer.
-	OnMessage func(datachannel.Payload)
-
-	// OnOpen designates an event handler which is invoked when
-	// the underlying data transport has been established (or re-established).
-	OnOpen func()
+	onMessageFn func(datachannel.Payload)
+	onOpenFn    func()
 
 	// Deprecated: Will be removed when networkManager is deprecated.
 	rtcPeerConnection *RTCPeerConnection
 }
 
+// OnOpen sets an event handler which is invoked when
+// the underlying data transport has been established (or re-established).
+func (d *RTCDataChannel) OnOpen(f func()) {
+	d.Lock()
+	defer d.Unlock()
+	d.onOpenFn = f
+}
+
+func (d *RTCDataChannel) onOpen() {
+	d.RLock()
+	defer d.RUnlock()
+	if d.onOpenFn == nil {
+		return
+	}
+
+	go func() {
+		d.onOpenFn()
```

Comment:
I think this is a mini-race since this is running in a separate routine the defer statement on the original routine may have already released the lock and someone else could have called ``OnOpen`` in the mean time. I see two options:
1. Store ``d.onOpenFn`` in a variable and pass it to the new routine. This leaves a tiny change of calling an outdated ``onOpenFn``. Seems rather acceptable thought.
2. Only acquire the lock in the new routine. This however means we'll always spawn a routine, even when the callback func is nil.

Reply:
Ah, yes. You're right, it is a race, and I like your first proposal to fix it.